src/memspaces/memspace_default.c

```diff
@@ -0,0 +1,91 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <assert.h>
+#include <hwloc.h>
+#include <stdlib.h>
+
+#include "base_alloc_global.h"
+#include "memory_target_numa.h"
+#include "memspace_internal.h"
+#include "memspace_numa.h"
+#include "utils_concurrency.h"
+
+static umf_result_t umfMemspaceDefaultCreate(umf_memspace_handle_t *hMemspace) {
+    if (!hMemspace) {
+        return UMF_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    umf_result_t umf_ret = UMF_RESULT_SUCCESS;
+
+    hwloc_topology_t topology;
+    if (hwloc_topology_init(&topology)) {
+        // TODO: What would be an approrpiate err?
+        return UMF_RESULT_ERROR_UNKNOWN;
+    }
+
+    if (hwloc_topology_load(topology)) {
+        umf_ret = UMF_RESULT_ERROR_UNKNOWN;
+        goto err_topology_destroy;
+    }
+
+    // Shouldn't return -1, as 'HWLOC_OBJ_NUMANODE' doesn't appear to be an
+    // object that can be present on multiple levels.
+    // Source: https://www.open-mpi.org/projects/hwloc/doc/hwloc-v2.10.0-letter.pdf
+    int nNodes = hwloc_get_nbobjs_by_type(topology, HWLOC_OBJ_NUMANODE);
+    assert(nNodes != -1);
+
+    size_t *nodeIds = umf_ba_global_alloc(nNodes * sizeof(size_t));
+    if (!nodeIds) {
+        umf_ret = UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+        goto err_topology_destroy;
+    }
+
+    // Collect all available NUMA node ids on the platform
+    int nodeIdx = 0;
+    hwloc_obj_t numaNodeObj = NULL;
+    while ((numaNodeObj = hwloc_get_next_obj_by_type(
+                topology, HWLOC_OBJ_NUMANODE, numaNodeObj)) != NULL) {
+        // Shouldn't be possible to iterate over more 'HWLOC_OBJ_NUMANODE' objects
+        // than the number returned by hwloc_get_nbobjs_by_type.
+        assert(nodeIdx < nNodes);
+        nodeIds[nodeIdx++] = numaNodeObj->os_index;
+    }
+
+    umf_ret =
+        umfMemspaceCreateFromNumaArray(nodeIds, (size_t)nNodes, hMemspace);
+
+    umf_ba_global_free(nodeIds, nNodes * sizeof(size_t));
+
+err_topology_destroy:
+    hwloc_topology_destroy(topology);
+    return umf_ret;
+}
+
+static umf_memspace_handle_t UMF_MEMSPACE_DEFAULT = NULL;
+static UTIL_ONCE_FLAG UMF_MEMSPACE_DEFAULT_INITIALIZED = UTIL_ONCE_FLAG_INIT;
+
+static void umfMemspaceDefaultDestroy(void) {
+    assert(UMF_MEMSPACE_DEFAULT);
+    umfMemspaceDestroy(UMF_MEMSPACE_DEFAULT);
+    UMF_MEMSPACE_DEFAULT = NULL;
+}
+
+static void umfMemspaceDefaultInit(void) {
+    umf_result_t ret = umfMemspaceDefaultCreate(&UMF_MEMSPACE_DEFAULT);
+    assert(ret == UMF_RESULT_SUCCESS);
+    (void)ret;
+
+    atexit(umfMemspaceDefaultDestroy);
```

Comment:
The problem is that there's no library destructor on windows when we build as a static library. I think we can do the same as for base_alloc: https://github.com/oneapi-src/unified-memory-framework/blob/main/src/base_alloc/base_alloc_global.c#L63 We use atexit when building as static lib on windows and call umf_ba_destroy_global in library dtor otherwise.

Reply:
In general, what Igor suggested makes sense if we want to support dynamic and static versions of UMF. 
But eventually building UMF core as a static library does not make much sense - tracking would not work (as expected).

So I agree with what @igchor suggested.

Reply:
Done. I've followed the example that @igchor gave. I've removed the `atexit` altogether since DEFAULT memspace depends on OS provider that is unsupported on Windows for now.